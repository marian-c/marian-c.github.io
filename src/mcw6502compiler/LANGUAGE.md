# MCW6502 language

This document describes the assembly language as supported by mcw6502

For actual documentation on how the instructions work, you can try:

- https://www.masswerk.at/6502/6502_instruction_set.html
- https://web.archive.org/web/20210426082142/http://www.obelisk.me.uk/6502/

## Whitespace and casing

- multi white space does not matter
- empty lines don't matter
- an empty program is a valid program
- casing does not matter: `BRK` or `brk` or `BrK` are equivalent

## Comments

Comments start with a `;`

```
; this is a comment
```

Comments can also be placed at the end of the line, after other stuff

```
myLabel: ; this is a comment
BRK ; this is a comment
```

## Valuables

We call valuable something that refers to a value, it can be a reference
(to a symbol or label) or a literal

### Reference to symbol

```
some_symbol = $1234 ; `$1234` is the valuable here
; ...
LDA some_symbol ; `some_symbol` is the valuable here
```

### Reference to label

```
location:
; ...
LDA: location
```

### Hex literal

```
LDA $1
LDA $01
LDA $1234
something = $12345
.org $0011
```

### Decimal literal

```
LDA 123
; plus all other examples from Hex literal
```

### Binary literal

```
LDA %1234
; plus all other examples from Hex literal
```

### Ascii char literal

```
LDA "a"
; plus all other examples from Hex literal
```

### Ascii string literal

```
LDA "ab"
; plus all other examples from Hex literal
```

## Label definitions:

Labels definitions are words followed my a `:` immediately

```
myLabel:
```

You can also put a comment on the same line, after the label

```
myLabel: ; describing this line
```

You can not have multiple labels with the same name;

You can use the label before it is defined.

## Symbol definitions

You can assign any valuable to a symbol.

```
my_symbol = $1234;
my_other_symbol = my_symbol;
```

Symbols can not be used before defining them.

## Instructions

### Instructions without operands

Some instructions are simple

```
ROR
```

### Instructions with operands

#### Absolute addressing mode (abs)

The operator can be any valuable except hex literals of length 1 or 2, must fit in 2 bytes

```
LDX $abcd
```

#### Absolute, X-indexed (abx)

The operator can be any valuable except hex literals of length 1 or 2, must fit in 2 bytes

```
LDX $abcd,X
```

#### Absolute, Y-indexed (aby)

The operator can be any valuable except hex literals of length 1 or 2, must fit in 2 bytes

```
LDX $abcd,Y
```

#### Immediate (imm)

The operator can be any valuable, must fit in 1 byte

```
LDA #$a1
```

#### Indirect (ind)

The operator can be any valuable, must fit in 2 bytes

```
LDA ($a1a2)
```

#### X-indexed, indirect (izx)

The operator can be any valuable, must fit in 1 byte

```
LDA ($a1,X)
```

#### Indirect, Y-indexed (izy)

The operator can be any valuable, must fit in 1 byte

```
LDA ($a1),Y
```

#### Relative (rel)

The operator can be any valuable, must fit in 2 bytes

```
BNE existing_label
```

#### Zeropage (zp0)

Either a hex literal with length 1 or 2

```
LDA $a1
```

or any valuable that fits in 1 byte, with forcing suffix

```
LDA.b my_symbol
```

#### Zeropage, X-indexed (zpx)

Either a hex literal with length 1 or 2

```
LDA $a1,x
```

or any valuable that fits in 1 byte, with forcing suffix

```
LDA.b my_symbol,x
```

#### Zeropage, Y-indexed (zpy)

Either a hex literal with length 1 or 2

```
LDA $a1,Y
```

or any valuable that fits in 1 byte, with forcing suffix

```
LDA.b my_symbol,Y
```

## Directives

### The "insert word" directive

The argument can be any valuable, must fit in 2 bytes

```
.word $1234
```

### The "set program counter" directive

> Fill value argument not implemented yet

The argument can be any valuable, must fit in 2 bytes

```
.org $1234
```

You can also set the value to fill the empty space

// TODO: implement the second argument

```
.org $1234 $ea
```

Compilation will fail if there are overwrites; it will also fail if the `.org` directives are not
properly ordered

Empty `.org` directives are possible; The next code fills the area from `$0800` to `$1000` with `$ea`

```
.org $0800
.org $1000 $ea
```

Having the second argument present (the fill value) will have no effect in certain situations, for example:

```
.org $0002 $ea ; fill value has no effect before the first byte
```

Be careful

```
myLbl: ; this is actually $0000, jumping here would make no sense in this example
.org $1000
lda $abcd ; this is $1000
```

## Notes

- labels and symbols are case-insensitive
- labels and symbols share the same namespace and can not be duplicated
- if an .org directive references a label, this is syntactically valid, but
  it would error because of overwrite
- you could reference a label from a variable, that's technically "use before define"
  but that's fine because labels are guaranteed to get resolved in the first pass,
  whereas variables do not have this guarantee
