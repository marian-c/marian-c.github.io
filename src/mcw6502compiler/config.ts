export type InstructionsInfo = Record<string, { list: string[]; opCodes: Record<string, number> }>;
export type FinalConfig = {
  __isFinal: true;
  // {LDA: {list: ['izx', 'izy'], opCodes: {IZX: 1, IZY: 2}}}
  instructionsInfo: InstructionsInfo;
};

type OptionalProps = never;

export type OptionalConfig = Omit<FinalConfig, OptionalProps | '__isFinal'> &
  Partial<Pick<FinalConfig, OptionalProps>>;

export type Config = OptionalConfig | FinalConfig;

export function makeConfig(initialConfig: Config): FinalConfig {
  if ('__isFinal' in initialConfig) {
    return initialConfig;
  }

  return {
    ...initialConfig,
    instructionsInfo: Object.fromEntries(
      Object.entries(initialConfig.instructionsInfo).map(([ins, modes]) => {
        return [
          ins.toLowerCase(),
          {
            list: modes.list.map((i) => i.toLowerCase()),
            opCodes: Object.fromEntries(
              Object.entries(modes.opCodes).map(([addressingMode, idx]) => {
                return [addressingMode.toLowerCase(), idx];
              }),
            ),
          },
        ];
      }),
    ),
    __isFinal: true,
  };
}
