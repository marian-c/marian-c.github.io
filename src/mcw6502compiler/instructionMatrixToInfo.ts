import type { Instruction } from '@/vendor-in/my-emulator/6502/olc6502_lookup_types';
import type { InstructionsInfo } from '@/mcw6502compiler/config';
import { sureAccess } from '@/helpers/objects';

export function phase2(phase1: Record<string, Record<string, number>>): InstructionsInfo {
  // TODO: error if relative and (absolute or zp0) mode are configured for the same instruction
  //  they are syntactically incompatible, although the 6502 does not fall into this situation,
  return Object.fromEntries(
    Object.entries(phase1).map(([instructionName, config]) => {
      return [instructionName, { opCodes: config, list: Object.keys(config) }];
    }),
  );
}

/**
 * given an instruction matrix, produce the instructionsInfo to be used by the compiler
 */
export const instructionMatrixToInfo = (matrix: Instruction[]): InstructionsInfo => {
  const phase1 = matrix.reduce(
    (acc, el, idx) => {
      const [instructionName, _, addressingModeFn] = el;
      const addresingMode = addressingModeFn.fName;
      if (acc[instructionName] === undefined) {
        acc[instructionName] = {};
      }
      sureAccess(acc, instructionName)[addresingMode] = idx;
      return acc;
    },
    {} as Record<string, Record<string, number>>,
  );

  return phase2(phase1);
};
