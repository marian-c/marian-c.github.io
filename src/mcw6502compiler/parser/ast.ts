// TODO: make these ints (remove the values)
export enum NodeKind {
  program = 'p',
  line = 'l',
  reference = 'r',
  instruction = 'i',
  directive = 'd',
  comment = 'c',
  valueLiteral = 'vl',
  labelDefinition = 'ld',
  symbolDefinition = 'sd',
  instructionOperator = 'or',
  instructionOperand = 'od',
  immAddressingMode = 'imm',
  absAddressingMode = 'abs',
  zp0AddressingMode = 'zp0',
  abxAddressingMode = 'abx',
  abyAddressingMode = 'aby',
  zpxAddressingMode = 'zpx',
  zpyAddressingMode = 'zpy',
  indAddressingMode = 'ind',
  izxAddressingMode = 'izx',
  izyAddressingMode = 'izy',
  relAddressingMode = 'rel',
}

export enum InstructionKind {
  withoutOperand,
  withOperand,
}

// DirectiveKind is also used as value, so don't change individual values
export enum DirectiveKind {
  org = 'org',
  word = 'word',
}

export enum ValueLiteralKind {
  ascii = 'ascii',
  bin = 'bin',
  dec = 'dec',
  hex = 'hex',
}

export interface NodeBase {
  readonly kind: NodeKind;
  readonly range: [number, number];
  readonly source?: string;
}
export interface InstructionOperatorNode extends NodeBase {
  readonly kind: NodeKind.instructionOperator;
  readonly value: string;
  readonly forcedTo?: 'b';
}

export interface ValueLiteralNode extends NodeBase {
  readonly kind: NodeKind.valueLiteral;
  readonly valueLiteralKind: ValueLiteralKind;
  readonly originalStringValue: string;
}
export interface ReferenceNode extends NodeBase {
  readonly kind: NodeKind.reference;
  readonly referencedName: string;
}

export type ValuableNode = ValueLiteralNode | ReferenceNode;

export interface ImmAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.immAddressingMode;
  readonly value: ValuableNode;
}

export interface AbsAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.absAddressingMode;
  readonly value: ValuableNode;
}

export interface Zp0AddressingModeNode extends NodeBase {
  readonly kind: NodeKind.zp0AddressingMode;
  readonly value: ValuableNode;
}

export interface AbxAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.abxAddressingMode;
  readonly value: ValuableNode;
}
export interface AbyAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.abyAddressingMode;
  readonly value: ValuableNode;
}
export interface ZpxAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.zpxAddressingMode;
  readonly value: ValuableNode;
}
export interface ZpyAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.zpyAddressingMode;
  readonly value: ValuableNode;
}
export interface IndAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.indAddressingMode;
  readonly value: ValuableNode;
}
export interface IzxAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.izxAddressingMode;
  readonly value: ValuableNode;
}
export interface IzyAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.izyAddressingMode;
  readonly value: ValuableNode;
}

export interface RelAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.relAddressingMode;
  readonly value: ValuableNode;
}

export type AddressingModeNode =
  | ImmAddressingModeNode
  | AbsAddressingModeNode
  | Zp0AddressingModeNode
  | AbxAddressingModeNode
  | AbyAddressingModeNode
  | ZpxAddressingModeNode
  | ZpyAddressingModeNode
  | IndAddressingModeNode
  | IzxAddressingModeNode
  | IzyAddressingModeNode
  | RelAddressingModeNode;

export type InstructionOperandNode = AddressingModeNode;

export interface CommentNode extends NodeBase {
  readonly kind: NodeKind.comment;
  readonly value: string;
}

export interface InstructionWithoutOperandNode extends NodeBase {
  readonly kind: NodeKind.instruction;
  readonly instructionKind: InstructionKind.withoutOperand;
  readonly operator: InstructionOperatorNode;
}
export interface InstructionWithOperandNode extends NodeBase {
  readonly kind: NodeKind.instruction;
  readonly instructionKind: InstructionKind.withOperand;
  readonly operator: InstructionOperatorNode;
  readonly operand: InstructionOperandNode;
}

export interface LabelDefinitionNode extends NodeBase {
  readonly kind: NodeKind.labelDefinition;
  readonly value: string;
}

export type SymbolDefinitionValueNode = ValueLiteralNode;

export interface SymbolDefinitionNode extends NodeBase {
  readonly kind: NodeKind.symbolDefinition;
  readonly symbolName: string;
  readonly symbolValue: SymbolDefinitionValueNode;
}

export interface DirectiveNodeBase extends NodeBase {
  readonly kind: NodeKind.directive;
  readonly directiveKind: DirectiveKind;
}

export interface DirectiveOrgNode extends DirectiveNodeBase {
  readonly directiveKind: DirectiveKind.org;
  // TODO: future: could allow .org directives to reference
  //   symbols if they are resolved by now
  readonly address: ValueLiteralNode;
  // TODO: (fill) actually implement this
  // readonly fill?: ValuableNode;
}
export interface DirectiveWordNode extends DirectiveNodeBase {
  readonly directiveKind: DirectiveKind.word;
  readonly value: ValuableNode;
}

export type DirectiveNode = DirectiveOrgNode | DirectiveWordNode;

export type InstructionNode = InstructionWithoutOperandNode | InstructionWithOperandNode;

export type LineContentsNode =
  | InstructionNode
  | LabelDefinitionNode
  | SymbolDefinitionNode
  | DirectiveNode;

export interface LineNode extends NodeBase {
  kind: NodeKind.line;
  comment?: CommentNode | undefined;
  contents?: LineContentsNode;
}

export interface ProgramNode extends NodeBase {
  readonly kind: NodeKind.program;
  readonly lines: readonly LineNode[];
}
