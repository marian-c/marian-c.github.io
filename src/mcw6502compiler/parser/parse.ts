import { type Config, makeConfig } from '@/mcw6502compiler/config';
import {
  type ErrorResult,
  makeSuccess,
  type ParseResult,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import { type LineContentsNode, type LineNode, NodeKind } from '@/mcw6502compiler/parser/ast';
import { parseComment } from '@/mcw6502compiler/parser/parseComment';
import {
  labelDefinitionRegExp,
  parseLabelDefinition,
} from '@/mcw6502compiler/parser/parseLabelDefinition';
import { parseInstruction } from '@/mcw6502compiler/parser/parseInstruction';
import { emptyRegExp, symbolDefinitionLightRegExp } from '@/mcw6502compiler/parser/regExp';
import { parseDirective } from '@/mcw6502compiler/parser/parseDirective';
import { parseSymbolDefinition } from '@/mcw6502compiler/parser/parseSymbolDefinition';
import { expected, testRegExp } from '@/mcw6502compiler/parser/stringInput';

/**
 * Given a string (a source file) and some config, it will return the ParseResult
 */
export function parseSource(originalSource: string, initialConfig: Config): ParseResult {
  const config = makeConfig(initialConfig);

  const sourceNormalized = originalSource
    .replace('\r\n', '\n')
    .replace('\r', '\n')
    .replace('\t', ' ');
  const lines = sourceNormalized.split('\n');
  const meta = { originalLines: lines };
  // TODO: does it work with different line endings?
  let index = 0;
  const lineNodes: LineNode[] = [];
  let errors: ErrorResult[] = [];
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed) {
      // is it a comment only line
      if (trimmed.startsWith(';')) {
        const commentResult = parseComment({ index, value: line, meta }, config);
        if (commentResult.kind === ResultKind.Error) {
          errors.push(commentResult);
        } else {
          lineNodes.push({
            kind: NodeKind.line,
            range: commentResult.value.range,
            comment: commentResult.value,
          });
        }
      } else {
        let result: Result<LineContentsNode>;
        // is it a label definition
        if (labelDefinitionRegExp.test(trimmed)) {
          result = parseLabelDefinition({ index, value: line, meta }, config);
        } else if (trimmed.startsWith('.')) {
          result = parseDirective({ index, value: line, meta }, config);
        } else if (symbolDefinitionLightRegExp.test(trimmed)) {
          result = parseSymbolDefinition({ index, value: line, meta }, config);
        } else {
          result = parseInstruction({ index, value: line, meta }, config);
        }
        const lineResultWithComment = (function addComment(
          r: Result<LineContentsNode>,
        ): Result<LineNode> {
          if (r.kind === ResultKind.Error) {
            return r;
          }
          let rest = r.rest;

          let commentResult;
          if (rest.value.trim().startsWith(';')) {
            let comment = parseComment(rest, config);
            if (comment.kind === ResultKind.Error) {
              return comment;
            }
            commentResult = comment;
            rest = commentResult.rest;
          }

          // if there's no comment, it should be empty
          if (!commentResult && !testRegExp(rest, emptyRegExp)) {
            return expected(rest, 'End of input');
          }

          return makeSuccess<LineNode>(
            {
              kind: NodeKind.line,
              range: [r.value.range[0], rest.index],
              contents: r.value,
              comment: commentResult?.value,
            },
            rest,
          );
        })(result);
        if (lineResultWithComment.kind === ResultKind.Error) {
          errors.push(lineResultWithComment);
        } else {
          lineNodes.push(lineResultWithComment.value);
        }
      }
    }

    index += line.length + 1;
  }

  return {
    errors,
    program: {
      kind: NodeKind.program,
      lines: lineNodes,
      range:
        lineNodes.length >= 1
          ? [lineNodes[0]!.range[0], lineNodes[lineNodes.length - 1]!.range[1]]
          : [0, sourceNormalized.length],
    },
    originalLines: lines,
  };
}
