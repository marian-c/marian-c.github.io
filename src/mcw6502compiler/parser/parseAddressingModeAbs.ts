import { expected, type StringInput, trimStart } from '@/mcw6502compiler/parser/stringInput';
import { type FinalConfig } from '@/mcw6502compiler/config';
import {
  makeError,
  makeSuccess,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import {
  type AbsAddressingModeNode,
  type InstructionOperatorNode,
  NodeKind,
  ValueLiteralKind,
} from '@/mcw6502compiler/parser/ast';
import { parseValuable } from '@/mcw6502compiler/parser/parseValuable';

export function parseAddressingModeAbs(
  stringInput: StringInput,
  config: FinalConfig,
  instruction: InstructionOperatorNode,
): Result<AbsAddressingModeNode> {
  const suffixValue = instruction.forcedTo;
  if (suffixValue === 'b') {
    return makeError(
      'ABS addressing mode can not be forced to byte',
      instruction.range,
      stringInput.meta.originalLines,
    );
  }
  const trimmed = trimStart(stringInput);

  const valuableResult = parseValuable(trimmed, config);
  if (valuableResult.kind === ResultKind.Error) {
    return valuableResult;
  }

  // abs syntax is a subset of abx or aby
  const rest = valuableResult.rest;
  if (rest.value.startsWith(',')) {
    return expected(rest, 'end of ABS mode');
  }

  const valuable = valuableResult.value;
  let hexLength = null;
  if (
    valuable.kind === NodeKind.valueLiteral &&
    valuable.valueLiteralKind === ValueLiteralKind.hex
  ) {
    hexLength = valuable.originalStringValue.length;
  }
  if (hexLength === 2 || hexLength === 1) {
    return makeError(
      'Hex literals must not have length 1 or 2 for ABS mode',
      [valuable.range[0] + 1, valuable.range[1]],
      stringInput.meta.originalLines,
    );
  }

  return makeSuccess<AbsAddressingModeNode>(
    {
      kind: NodeKind.absAddressingMode,
      value: valuableResult.value,
      range: [trimmed.index, rest.index],
    },
    valuableResult.rest,
  );
}
