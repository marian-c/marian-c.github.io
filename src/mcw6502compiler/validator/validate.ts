import {
  type ErrorResult,
  makeError,
  type ParseResult,
} from '@/mcw6502compiler/parser/parserTypes';
import { NodeKind } from '@/mcw6502compiler/parser/ast';
import type { Config } from '@/mcw6502compiler/config';

/**
 * if the returned value has length 0, there are no errors
 */
export function validateParseTree(parseTree: ParseResult, _initialConfig: Config): ErrorResult[] {
  const errors: ErrorResult[] = [];
  const identifiers = new Set<string>();

  // check for duplicate identifiers
  for (const line of parseTree.program.lines) {
    let identifier = null;
    if (line.contents?.kind === NodeKind.labelDefinition) {
      identifier = line.contents.value;
    }

    if (line.contents?.kind === NodeKind.symbolDefinition) {
      identifier = line.contents.symbolName;
    }

    // TODO: better error location
    if (identifier !== null) {
      if (identifiers.has(identifier)) {
        errors.push(
          makeError(`Duplicate identifier: ${identifier}`, line.range, parseTree.originalLines),
        );
      }
      identifiers.add(identifier);
    }
  }

  // TODO: validate missing references

  return errors;
}
