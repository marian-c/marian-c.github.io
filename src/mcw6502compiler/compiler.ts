import { parseSource } from '@/mcw6502compiler/parser/parse';
import { type Config, makeConfig } from '@/mcw6502compiler/config';
import { validateParseTree } from '@/mcw6502compiler/validator/validate';
import { assemble } from '@/mcw6502compiler/assembler/assemble';
import type { ErrorResult } from '@/mcw6502compiler/parser/parserTypes';
import type { ExtractSuccessType, GenericResult } from '@/types';
import { GenericResultKind } from '@/types';

export function compile(
  originalSource: string,
  initialConfig: Config,
): GenericResult<ErrorResult, ExtractSuccessType<ReturnType<typeof assemble>>> {
  const config = makeConfig(initialConfig);
  let errors: ErrorResult[] = [];

  const parseResult = parseSource(originalSource, config);
  errors.push(...parseResult.errors);

  if (!errors.length) {
    errors.push(...validateParseTree(parseResult, config));
  }

  if (!errors.length) {
    return assemble(parseResult, config);
  }

  return { kind: GenericResultKind.error, errors };
}
