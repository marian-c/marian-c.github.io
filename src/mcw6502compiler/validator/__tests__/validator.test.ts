import { parseSource } from '@/mcw6502compiler/parser/parse';
import { validateParseTree } from '@/mcw6502compiler/validator/validate';
import { phase2 } from '@/mcw6502compiler/instructionMatrixToInfo';

const instructionsInfo = phase2({
  BRK: { IMP: 1 },
  BNE: { REL: 2 },
});

function handle(programBody: string) {
  const config = { instructionsInfo };
  const programResult = parseSource(programBody, config);
  const validationResult = validateParseTree(programResult, config);
  return { programResult, validationResult };
}

describe('Validator', () => {
  test('Positive - empty', () => {
    const { programResult, validationResult } = handle(``);
    expect(programResult.errors).toHaveLength(0);
    expect(validationResult).toMatchSnapshot();
  });

  test('Positive - label defined first', () => {
    const { programResult, validationResult } = handle(`
    start:
    bne start
    `);
    expect(programResult.errors).toHaveLength(0);
    expect(validationResult).toMatchSnapshot();
  });

  test('Positive - label referenced first', () => {
    const { programResult, validationResult } = handle(`
bne start
start:
    `);
    expect(programResult.errors).toHaveLength(0);
    expect(validationResult).toMatchSnapshot();
  });

  test('Error when duplicate labels', () => {
    const { programResult, validationResult } = handle(`
start:
start: ; same label
    `);
    expect(programResult.errors).toHaveLength(0);
    expect(validationResult).toMatchSnapshot();
  });

  test('Error when duplicate symbols', () => {
    const { programResult, validationResult } = handle(`
my_symbol = 1
my_symbol = 2
    `);
    expect(programResult.errors).toHaveLength(0);
    expect(validationResult).toMatchSnapshot();
  });
  test('Error when duplicate symbols + labels', () => {
    const { programResult, validationResult } = handle(`
my_symbol = 1
my_symbol: 
BRK
    `);
    expect(programResult.errors).toHaveLength(0);
    expect(validationResult).toMatchSnapshot();
  });

  test('no error when label does not exist (handled by the assembler)', () => {
    const { programResult, validationResult } = handle(`
bne missing
existing:
    `);
    expect(programResult.errors).toHaveLength(0);
    expect(validationResult).toMatchSnapshot();
  });
});
