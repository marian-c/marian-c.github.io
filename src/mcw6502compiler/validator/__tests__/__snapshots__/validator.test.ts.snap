// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator Error when duplicate labels 1`] = `
[
  {
    "kind": 1,
    "lineRange": {
      "end": [
        3,
        20,
      ],
      "start": [
        3,
        1,
      ],
    },
    "message": "Duplicate identifier: start",
    "range": [
      8,
      27,
    ],
  },
]
`;

exports[`Validator Error when duplicate symbols + labels 1`] = `
[
  {
    "kind": 1,
    "lineRange": {
      "end": [
        3,
        11,
      ],
      "start": [
        3,
        1,
      ],
    },
    "message": "Duplicate identifier: my_symbol",
    "range": [
      15,
      25,
    ],
  },
]
`;

exports[`Validator Error when duplicate symbols 1`] = `
[
  {
    "kind": 1,
    "lineRange": {
      "end": [
        3,
        14,
      ],
      "start": [
        3,
        1,
      ],
    },
    "message": "Duplicate identifier: my_symbol",
    "range": [
      15,
      28,
    ],
  },
]
`;

exports[`Validator Positive - empty 1`] = `[]`;

exports[`Validator Positive - label defined first 1`] = `[]`;

exports[`Validator Positive - label referenced first 1`] = `[]`;

exports[`Validator no error when label does not exist (handled by the assembler) 1`] = `[]`;
